### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by Qi77Qi**
Ticket: No ticket
* Most changes are from https://github.com/broadinstitute/sam/pull/230.
Since 230 is blocked currently and lots of the changes were there so that it's easier to add lock, this PR removes most lock related changes and try to get non-lock related code into master first so that other PRs can benefit from the changes.
* address https://github.com/broadinstitute/sam/pull/233/files#diff-0e8c5e3352da69b276f3838bbca624beR90

The only lock related change in this PR is configuration, which I didn't bother remove since it won't affect any real code and we will add them in the lock PR anyway

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

### CODE REVIEW

**Review by marctalbott — APPROVED** [review]

File: src/main/resources/reference.conf [hunk 1]
```diff
@@ -359,3 +359,9 @@ resourceTypes = {
 
   }
 }
+
+
+distributedLock{
```
Comment (helgridly) [inline src/main/resources/reference.conf]: nit: space

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala [hunk 2]
```diff
@@ -30,121 +29,188 @@ import org.broadinstitute.dsde.workbench.sam.service._
...
-        val googlePubSubDAO = new HttpGooglePubSubDAO(googleServicesConfig.appName, Pem(WorkbenchEmail(googleServicesConfig.serviceAccountClientId), new File(googleServicesConfig.pemFile)), "google", googleServicesConfig.groupSyncPubSubProject)
-        val googleStorageDAO = new HttpGoogleStorageDAO(googleServicesConfig.appName, Pem(WorkbenchEmail(googleServicesConfig.serviceAccountClientId), new File(googleServicesConfig.pemFile)), "google")
-        val googleProjectDAO = new HttpGoogleProjectDAO(googleServicesConfig.appName, Pem(googleServicesConfig.billingPemEmail, new File(googleServicesConfig.pathToBillingPem), Option(googleServicesConfig.billingEmail)), "google")
-        val googleKeyCache = new GoogleKeyCache(googleIamDAO, googleStorageDAO, googlePubSubDAO, googleServicesConfig, petServiceAccountConfig)
-        val notificationDAO = new PubSubNotificationDAO(googlePubSubDAO, googleServicesConfig.notificationTopic)
+        _ <- IO.fromFuture(IO(dependencies.cloudExtensions.onBoot(dependencies.samApplication)))
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala]: `IO.fromFuture(IO(` :vomit:

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala [hunk 3]
```diff
@@ -228,44 +228,41 @@ class GoogleExtensions(val directoryDAO: DirectoryDAO, val accessPolicyDAO: Acce
...
-        case (Some(p), Some(_)) => Future.successful(p)
+        case (Some(p), Some(_)) => IO.pure(p)
       }
     } yield pet
 
-    // In the case of high concurrency, the above code may find that there is no pet SA and then try to create one, getting
-    // a 409 Conflict in Google if another call has since created it. This recoverWith will catch that case and return the
-    // newly-created pet SA instead of simply failing.
-    pet recoverWith {
+    createPet.unsafeToFuture() recoverWith {
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala:265]: keep the comment!

File: src/main/resources/reference.conf (hunk 1 above)
Comment (helgridly) [reply src/main/resources/reference.conf]: if you're gonna go down the scalafmt rabbithole here, i think what we're after is: `class Foo(` and `foo.map(` but `Foo {` and `blah map {`, i.e. space before braces but not before parens

**Review by helgridly — APPROVED** [review]
i've mainly checked for style here, rather than correctness

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala (hunk 3 above)
Comment (Qi77Qi) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala:265]: It'll be removed in the lock PR. so I'll leave this as is

**Review by dvoet — APPROVED** [review]

**Comment by Qi77Qi** [comment]
jenkins retest

**Comment by Qi77Qi** [comment]
jenkins retest

**Pull request merged**